### PR TITLE
Add MKLDNN fuser

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52614 Avoid dispatch overhead in call to mkldnn convolution
* #52613 Include max pool in fusion groups
* #52513 Extend subgraph utils to cover merging a node following a subgraph
* #52512 Remove unused subgraph vmap api
* #51924 Only functionalize ops which we want to include in mkldnn group
* #51923 Add a filter to remove mutation
* #51922 Dont error on 0-dim in convolution
* #51921 Add 0-dim support for binary mkldnn ops
* #51736 Handle mkldnn broadcasting in mkldnn fuser
* **#51600 Add MKLDNN fuser**
* #52786 [NNC] Make NNC sanitize input names
* #51484 Add support for linear in mkldnn fusion
* #51483 Move conv to mkldnn

As the benchmarks in #51484 and #51483 show, pre-converting the weights of Convolution and Linear can lead to significant speedups. In this simple version we're essentially doing what exists in `torch/utils/mkldnn` for convolution. We can convert the input to mkldnn, and the output back to dense. However, we can improve the performance form this naive version by 
avoiding conversions to and from mkldnn. We do this by finding subgraphs of nodes computable in MKLDNN - this is essentially ops with `MkldnnCPU` dispatch keys in `native_functions.yaml`. 

As stated in the comments for the fuser:
```
// [mkldnn perf strategy]
// Certain ops - aten::linear, aten::conv2d, aten::conv3d - provide a huge speed
// up just by converting the constant weights to MKLDNN AOT, and then at runtime
// converting the non-constant input to_mkldnn before the op, and then back to
// its original layout after the op. The speed up holds even if you end up
// converting the input to_mkldnn and output back to_dense. We start groups of
// ops to compute in MKLDNN only from these ops that are a strict speedup. Then,
// we expand the groups to include operators which are computable in MKLDNN &
// are roughly perf equal to eager. We do this in the hopes of joining multiple
// fast nodes together, saving to_mkldnn and to_dense conversions.
//
// MKLDNN only supports float32 inputs for aten::linear, aten::conv2d &
// aten::conv3d. We only fuse these nodes if the weights are float32, and then
// we only include operators which we can prove will execute in float32. By
// fusing topologically we can maintain the invariant that all tensor types in
// the graph are floating point. In fusing Conv-> Add -> Relu -> Conv we start
// with the first Conv, know that the output is float, and can then safely merge
// Add and Relu. If we started with the last Conv, it would be difficult to
// prove in our first pass that the Add's inputs were both float32 without first
// fusing the first conv.
```

Differential Revision: [D26696702](https://our.internmc.facebook.com/intern/diff/D26696702)

### CODE REVIEW

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 1]
```diff
@@ -165,97 +205,262 @@ bool frozenMkldnnCompatibleConvNode(Node* n) {
...
 } // namespace
 
 void ConvertFrozenOpsToMKLDNN(std::shared_ptr<Graph>& graph) {
 #if AT_MKLDNN_ENABLED()
+  GRAPH_DUMP("Before convert frozen ops to mkldnn", graph);
+  // TODO: replace conv1d with conv2d ?
   graph_rewrite_helper::replaceConvolutionWithAtenConv(graph);
-  ConvertFrozenConvParamsToMKLDNN(graph->block());
-  EliminateDeadCode(graph);
+  if (containsMKLDNNGroup(graph->block())) {
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:449]: I wonder if it might also make sense to have another check similar `tensorInputIsMKLDNNSupported` to make a bit more conservative?

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 2]
```diff
@@ -101,32 +60,113 @@ Node* createConstantMKLDNNTensorOp(Graph* g, Tensor mkldnn_tensor) {
 }
 
 bool supportedMKLDNNWeight(const Tensor& weight) {
-  if (!weight.device().is_cpu()) {
-    return false;
+  return weight.device().is_cpu() && weight.dtype() == c10::ScalarType::Float &&
+      weight.ndimension() != 0;
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:64]: I guess it won't make much sense for inputs to be 0-dim

**Review by Krovatkin — APPROVED** [review]
a few comments.

File: aten/src/ATen/core/Formatting.cpp [hunk 3]
```diff
@@ -241,6 +241,9 @@ std::ostream& print(std::ostream& stream, const Tensor & tensor_, int64_t linesi
     Tensor tensor;
     if (tensor_.is_quantized()) {
       tensor = tensor_.dequantize().to(kCPU, kDouble).contiguous();
+    } else if (tensor_.is_mkldnn()) {
+      stream << "MKLDNN Tensor: ";
+      tensor = tensor_.to_dense().to(kCPU, kDouble).contiguous();
```
[the remark this answers is not in the document]
Comment (eellison) [reply aten/src/ATen/core/Formatting.cpp:246]: Yea. A lot of the formatting assumes cpu double :/ If you see quantized tensors do the same thing as im doing here 

`tensor_.dequantize().to(kCPU, kDouble).contiguous()`

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:64]: Actually, this restriction might be able to removed, since i'm adding for support for 0-dim tensors in a later PR.

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp [hunk 4]
```diff
@@ -52,6 +52,12 @@ struct ValueMapper {
         subgraph_num_outputs_, num_outputs - subgraph_num_outputs_);
     for (Value* v : new_outputs) {
       auto maybe_last_use = firstOrLastUse(v, /*find_first*/ false);
+      if (!maybe_last_use) {
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:55]: i think so, it may also because we're fusing in the wrong order and i'm doing this:
```
    // TODO: subgraph utils doesnt handle merging a consumer into a producer
    // so we merge the producer into the consumer even though it's inefficient
    consumer = SubgraphUtils::createSingletonSubgraphAndUpdateAliasing(
```

I'm going to try to address that in a follow up PR. It makes the fuser slower, but this is run once AOT once so being a little slower is not a blocking issue IMO

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:449]: That check is covered as part of this call

File: test/jit/test_freezing.py [hunk 5]
```diff
@@ -1589,21 +1578,54 @@ def forward(self, input):
...
+                self.run_pass("convert_frozen_ops_to_mkldnn", mod.graph)
+                FileCheck().check("to_mkldnn").check("linear").check("to_dense").check("add").run(mod.graph)
+
+            lin = LinearMod(20, 20)
+            # Scalar-Tensor not supported
+            test_unsupported(nn.Sequential(lin, Add(.5)))
+            # # 0-dim not supported
+            test_unsupported(nn.Sequential(lin, Add(torch.tensor(.5))))
+            # tensor of unknown dtype (getAttr node here) not supported
+            test_unsupported(nn.Sequential(lin, Add(torch.tensor([20]))), ['1'])
```
[the remark this answers is not in the document]
Comment (eellison) [reply test/jit/test_freezing.py:1628]: There is actually one: `mod = nn.Sequential(LinearMod(20, 20), LinearMod(20, 20)).eval()`, i can add a few more..

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:55]: this does get addressed in a later PR
